design-proposals/edge-infra-day2-improvements.md

```diff
@@ -0,0 +1,485 @@
+        MM->>Inventory: Update "os_updates_available" field on the Instance with the new EMT version
+        MM->>PUA: PlatformUpdateResponse
+    else Mutable (Ubuntu) OS
+        note over MM,PUA: New package updates are detected by PUA
+        PUA->>MM: PlatformUpdateRequest with new package versions
+        MM->>Inventory: Update "os_updates_available" field on the Instance with the list of new package versions available
+        MM->>PUA: PlatformUpdateResponse
+    end
+    end
+```
+
+Installed packages filled in OSProfile and runtime packages in Instances:
+```mermaid
+sequenceDiagram
+%%{wrap}%%
+autonumber
+    box Edge Infrastructure Manager
+    participant Inventory
+    participant OSRM as OS Resource Manager
+    participant uRepo as Ubuntu repository
+    participant MM as Maintenance Manager
+    end
+    box Edge Node
+    participant PUA as Platform Update Agent
+    end
+    
+    
+    loop OSRM creates the OSProfile for Mutable OS (Ubuntu)
+        OSRM->>uRepo: Fetch the Manifest from the Ubuntu repository
+        OSRM->>OSRM: Parses the Ubuntu Manifest file
+        OSRM->>Inventory: Packages info into the "installed_packages" field in the OSProfile
+    end
+    
+    loop Edge Node checking for installed packages on the OS
+        MM->>PUA: List of installed packages
+        PUA->>Inventory: Compare and update "runtime_packages" field in the Instance if needed
+    end
+```
+
+### Orchestrator Update and Migrations
+
+Given the above schema changes, we need to consider how we are going to support
+orchestrator updates and data migrations in the database.
+
+**Mutable OS** 
+- Any Mutable OSProfile with `update_sources`, `kernel_commands` or
+  `installed_packages` fields set, needs to be migrated to the new schema by
+  creating OSUpdatePolicy, filling the `install_packages`, `update_sources` and
+  `kernel_command` fields with the one coming from the OSprofile and linking it
+  to the related Instances. 
+- In any Mutable OSProfile the `update_sources`, `kernel_commands` and
+  `installed_packages` fields must be cleaned, since the fields are either
+  deprecated or change their meaning. This should happen after the creation of
+  the OSUpdatePolicy.
+
+> Note: The `installed_packages` field for Mutable OSProfile won't be populated
+> for OSProfile migrated from old orchestrator versions.
+
+**Immutable OSes**
+  - For any Instance with a `desired_os` set that is different from the
+    `current_os`, we must create a OSUpdatePolicy of type `POLICY_TARGET` with
+    `target_os` equal to the `desired_os` and link it to the Instance.
+
+We don't expect to create OSUpdateRun for past updates, since we don't have
+information about what happened during updates in the past.
+
+New fields in the Instance resource, such as `runtime_packages`,
+`os_update_available`, won't be populated during migration, but will be
+eventually populated by the Maintenance Manager.
+
+### EMT versioning and automatic creation of OSProfiles
```

Comment:
shall we consider a mechanism to invalidate OS profiles created by the USER to prevent any user-driven error? I am also thinking that looking at specific component versions (like we discussed Yesterday) could be a safer approach just than looking at the os profile version

Reply:
I see your point, and I agree. Users could create issues if they create overlapping OSProfiles. However, as I wrote in the document, the manual addition of OSProfiles is an advanced feature. I don't think checking agents version would solve this.